src/stores/right-panel/RightPanelStore.ts

```diff
@@ -0,0 +1,339 @@
+        const redirect = this.getVerificationRedirect(card);
+        const targetPhase = redirect?.phase ?? card.phase;
+        const cardState = redirect?.state ?? (Object.keys(card.state ?? {}).length === 0 ? null : card.state);
+
+        // Checks for wrong SetRightPanelPhase requests
+        if (!this.isPhaseActionIsValid(targetPhase)) return;
+
+        if (targetPhase === this.currentCard?.phase && allowClose && !cardState) {
+            // Toggle panel: a toggle command needs to fullfil the following:
+            // - the same phase
+            // - the panel can be closed
+            // - does not contain any state information (state)
+            this.togglePanel(rId);
+            return;
+        } else if ((targetPhase === this.currentCardForRoom(rId)?.phase && !!cardState)) {
+            // Update state: set right panel with a new state but keep the phase (dont know it this is ever needed...)
+            const hist = this.byRoom[rId].history ?? [];
+            hist[hist.length - 1].state = cardState;
+            this.emitAndUpdateSettings();
+            return;
+        } else if (targetPhase !== this.currentCard?.phase) {
+            // Set right panel and erase history.
+            this.setRightPanelCache({ phase: targetPhase, state: cardState ?? {} });
+        }
+    }
+
+    public pushCard(
+        card: IRightPanelCard,
+        allowClose = true,
+        roomId: string = null,
+    ) {
+        const rId = roomId ?? this.viewedRoomId;
+        const redirect = this.getVerificationRedirect(card);
+        const targetPhase = redirect?.phase ?? card.phase;
+        const pState = redirect?.state ?? (Object.keys(card.state ?? {}).length === 0 ? null : card.state);
+
+        // Checks for wrong SetRightPanelPhase requests
+        if (!this.isPhaseActionIsValid(targetPhase)) return;
+
+        let roomCache = this.byRoom[rId];
+        if (!!roomCache) {
+            // append new phase
+            roomCache.history.push({ state: pState, phase: targetPhase });
+            roomCache.isOpen = allowClose ? roomCache.isOpen : true;
+        } else {
+            // create new phase
+            roomCache = {
+                history: [{ phase: targetPhase, state: pState ?? {} }],
+                // if there was no right panel store object the the panel was closed -> keep it closed, except if allowClose==false
+                isOpen: !allowClose,
+            };
+        }
+
+        this.emitAndUpdateSettings();
+    }
+
+    public popCard(roomId: string = null) {
+        const rId = roomId ?? this.viewedRoomId;
+        const roomCache = this.byRoom[rId];
+        const removedCard = roomCache.history.pop();
+        this.emitAndUpdateSettings();
+        return removedCard;
+    }
+
+    public togglePanel(roomId: string = null) {
+        const rId = roomId ?? this.viewedRoomId;
+        this.byRoom[rId].isOpen = !this.byRoom[rId].isOpen;
+        this.emitAndUpdateSettings();
+    }
+
+    // Private
```

Comment:
The other section headers in this file have no line break after, so choose one style that applies to all of them.